src/ccs/r1cs.rs

```diff
@@ -9,17 +10,65 @@ pub struct R1CS<F: PrimeField> {
     pub B: SparseMatrix<F>,
     pub C: SparseMatrix<F>,
 }
+
 impl<F: PrimeField> R1CS<F> {
     /// returns a tuple containing (w, x) (witness and public inputs respectively)
     pub fn split_z(&self, z: &[F]) -> (Vec<F>, Vec<F>) {
         (z[self.l + 1..].to_vec(), z[1..self.l + 1].to_vec())
     }
+
+    /// check that a R1CS structure is satisfied by a z vector. Only for testing.
+    pub fn check_relation(&self, z: &[F]) -> Result<(), Error> {
+        let Az = mat_vec_mul_sparse(&self.A, z);
+        let Bz = mat_vec_mul_sparse(&self.B, z);
+        let Cz = mat_vec_mul_sparse(&self.C, z);
+        let AzBz = hadamard(&Az, &Bz);
+        assert_eq!(AzBz, Cz);
+
+        Ok(())
+    }
+
+    pub fn relax(self) -> RelaxedR1CS<F> {
+        RelaxedR1CS::<F> {
+            l: self.l,
+            E: vec![F::zero(); self.A.n_rows],
+            A: self.A,
+            B: self.B,
+            C: self.C,
+            u: F::one(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct RelaxedR1CS<F: PrimeField> {
+    pub l: usize, // io len
+    pub A: SparseMatrix<F>,
+    pub B: SparseMatrix<F>,
+    pub C: SparseMatrix<F>,
+    pub u: F,
+    pub E: Vec<F>,
```

Comment:
Does all of this need to be `pub`?? Or can we make it `pub(crate)`?

Reply:
So, since it happened to me in other libraries that fields that are intended to not be public there is some other project that wants to use them, thus they end up making a PR to make them public, and also it happened to me from the other side, that I used some repo and suffered from not being able to access to some private fields, in cases like this (RelaxedR1CS structure, I put the parameters as public, so other people building stuff from here can use it as they wish.
If we were building a very strickt & safe library that only allows a very clear use case and we want to prevent missusage because the main usage is very clear, I would prefer to have those parameters as private, but for this folding schemes library I expect people to build stuff around it, even stuff that we can not imagine yet ^^